File: typescript/src/documents.ts
import type { HttpClient } from './http.js';
import type { DocumentsIndexResponse, TachyonDocument } from './types.js';

/** `/collections/{name}/documents` — index, fetch, and delete documents. */
export class DocumentsApi<T extends TachyonDocument = TachyonDocument> {
  constructor(
    private readonly http: HttpClient,
    private readonly collectionName: string,
  ) {}

  /**
   * `POST /collections/{name}/documents`. Accepts one document or an array;
   * documents are upserted by `id`. Individual documents can fail without
   * failing their neighbours — check `num_failed` and `results`.
   */
  index(documents: T | T[]): Promise<DocumentsIndexResponse> {
    return this.http.request<DocumentsIndexResponse>(
      'POST',
      `/collections/${encodeURIComponent(this.collectionName)}/documents`,
      { body: documents },
    );
  }

  /** `GET /collections/{name}/documents/{id}`. */
  retrieve(id: string): Promise<T> {
    return this.http.request<T>(
      'GET',
      `/collections/${encodeURIComponent(this.collectionName)}/documents/${encodeURIComponent(id)}`,
    );
  }

  /** `DELETE /collections/{name}/documents/{id}`. */
  async delete(id: string): Promise<void> {
    await this.http.request<void>(
      'DELETE',
      `/collections/${encodeURIComponent(this.collectionName)}/documents/${encodeURIComponent(id)}`,
    );
  }
}

File: go/documents.go
package tachyon

import (
	"context"
	"net/http"
	"net/url"
)

// DocumentsService is /collections/{name}/documents — index, fetch, and delete documents.
type DocumentsService struct {
	transport      *httpTransport
	collectionName string
}

// Index sends POST /collections/{name}/documents, upserting one or more
// documents by id. Individual documents can fail without failing their
// neighbours — check NumFailed and Results on the response.
func (s *DocumentsService) Index(ctx context.Context, docs ...Document) (DocumentsIndexResponse, error) {
	var out DocumentsIndexResponse
	path := "/collections/" + url.PathEscape(s.collectionName) + "/documents"
	err := s.transport.do(ctx, http.MethodPost, path, requestOptions{body: docs}, &out)
	return out, err
}

// Retrieve sends GET /collections/{name}/documents/{id}.
func (s *DocumentsService) Retrieve(ctx context.Context, id string) (Document, error) {
	var out Document
	path := "/collections/" + url.PathEscape(s.collectionName) + "/documents/" + url.PathEscape(id)
	err := s.transport.do(ctx, http.MethodGet, path, requestOptions{}, &out)
	return out, err
}

// Delete sends DELETE /collections/{name}/documents/{id}.
func (s *DocumentsService) Delete(ctx context.Context, id string) error {
	path := "/collections/" + url.PathEscape(s.collectionName) + "/documents/" + url.PathEscape(id)
	return s.transport.do(ctx, http.MethodDelete, path, requestOptions{}, nil)
}
